### PR TITLE
Hw solution jv fr UI t

### PR DESCRIPTION
**Pull request by DPotapov93**

### CODE REVIEW

File: src/main/java/core/basesyntax/Main.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        List<FruitTransaction> fruitTransactions = fruitMapper.convertData(allLines);
+        Map<FruitTransaction.Operation, OperationHandler> handlers = createOperationHandlers();
+        OperationStrategy operationStrategy = new OperationStrategy(handlers);
+        OperationProcess operationProcess = new OperationProcessImpl(operationStrategy);
+        operationProcess.processTransaction(fruitTransactions);
+        CreateReport createReport = new CreateReportImpl();
+        String report = createReport.writeReport(Storage.shopStorage);
+        WriterService writerService = new WriterServiceImp();
+        boolean isSuccessWrite = writerService.writeCvsFile(report, FILE_TO_WRITE);
+        System.out.println(isSuccessWrite);
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/Main.java:41]: Can we avoid it?

File: src/main/java/core/basesyntax/model/FruitTransaction.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    public int getQuantity() {
+        return quantity;
+    }
+
+    public void setQuantity(int quantity) {
+        this.quantity = quantity;
+    }
+
+    public enum Operation {
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/model/FruitTransaction.java:33]: Let's make separate file for it.

File: src/main/java/core/basesyntax/model/FruitTransaction.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        public static Operation getOperationByCode(String code) {
+            Operation[] values = Operation.values();
+            for (Operation operation : values) {
+                if (operation.code.equals(code)) {
+                    return operation;
+                }
+            }
+            throw new RuntimeException(EXCEPTION_MESSAGE);
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/model/FruitTransaction.java:51]: Let's use a narrower exception, more informative. 
IllegalArgumentException for example.

File: src/main/java/core/basesyntax/service/CreateReport.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import java.util.Map;
+
+public interface CreateReport {
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/service/CreateReport.java:5]: Rename.
Use nouns here - ReportCreator.

File: src/main/java/core/basesyntax/service/CreateReport.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax.service;
+
+import java.util.Map;
+
+public interface CreateReport {
+    String writeReport(Map<String, Integer> resume);
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/service/CreateReport.java:6]: If it is creator, let's create, don't write.

File: src/main/java/core/basesyntax/service/impl/CreateReportImpl.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    private static final String COMA = ",";
+
+    @Override
+    public String writeReport(Map<String, Integer> resume) {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append(TITLE).append(LINE_BREAK);
+        for (Map.Entry<String, Integer> entry : resume.entrySet()) {
+            String fruit = entry.getKey();
+            int amount = entry.getValue();
+            stringBuilder.append(fruit).append(COMA).append(amount).append(LINE_BREAK);
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/service/impl/CreateReportImpl.java:18]:
```suggestion
            stringBuilder.append(entry.getKey()).append(COMA).append(entry.getValue()).append(LINE_BREAK);

```

File: src/main/java/core/basesyntax/service/FruitMapper.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package core.basesyntax.service;
+
+import core.basesyntax.model.FruitTransaction;
+import java.util.List;
+
+public interface FruitMapper {
+    List<FruitTransaction> convertData(List<String> dataFromFile);
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/service/FruitMapper.java:7]: Rename.

File: src/main/java/core/basesyntax/strategy/impl/PurchaseOperationHandler.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    private static final String EXCEPTION_MESSAGE = "You don't have such fruits!";
+
+    @Override
+    public void handle(FruitTransaction fruitTransaction) {
+        String fruitName = fruitTransaction.getFruit();
+        Integer transactionQuantity = fruitTransaction.getQuantity();
+        if (Storage.shopStorage.containsKey(fruitName)) {
+            Integer availableQuantity = Storage.shopStorage.get(fruitName);
+            if (availableQuantity < transactionQuantity) {
+                throw new RuntimeException(LESS_QUANTITY_MESSAGE + fruitName);
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/strategy/impl/PurchaseOperationHandler.java:19]: Let's use a narrower exception here.

File: src/main/java/core/basesyntax/strategy/impl/PurchaseOperationHandler.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+            } else {
+                Storage.shopStorage.put(fruitName, availableQuantity - transactionQuantity);
+            }
+        } else {
+            throw new RuntimeException(EXCEPTION_MESSAGE);
```
Comment (bhdnchui) [inline src/main/java/core/basesyntax/strategy/impl/PurchaseOperationHandler.java:24]: And here.

**Review by bhdnchui — CHANGES_REQUESTED** [review]
Great job! Just a few comments need to be processed.

File: src/main/java/core/basesyntax/service/impl/OperationProcessImpl.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package core.basesyntax.service.impl;
+
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.service.OperationProcess;
+import core.basesyntax.strategy.OperationStrategy;
+import java.util.List;
+
+public class OperationProcessImpl implements OperationProcess {
+    private OperationStrategy operationStrategy;
```
Comment (kshuryhin) [inline src/main/java/core/basesyntax/service/impl/OperationProcessImpl.java:9]: Field `operationStrategy` may be `final`

File: src/main/java/core/basesyntax/service/impl/ReaderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    private static final String DO_NOT_EXIST_MESSAGE = "File does not exist: ";
+
+    @Override
+    public List<String> read(String filePath) {
+        if (filePath == null) {
+            throw new RuntimeException(NULL_MESSAGE);
+        }
+        File file = new File(filePath);
+        if (!file.exists()) {
+            throw new RuntimeException(DO_NOT_EXIST_MESSAGE + filePath);
```
Comment (kshuryhin) [inline src/main/java/core/basesyntax/service/impl/ReaderServiceImpl.java:23]: I think we can use more specific exception type in these situations

File: src/main/java/core/basesyntax/service/impl/ReportCreatorImpl.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package core.basesyntax.service.impl;
+
+import core.basesyntax.service.ReportCreator;
+import java.util.Map;
+
+public class ReportCreatorImpl implements ReportCreator {
+    private static final String TITLE = "fruit,quantity";
+    private static final String LINE_BREAK = "\n";
```
Comment (kshuryhin) [inline src/main/java/core/basesyntax/service/impl/ReportCreatorImpl.java:8]: i would recommend to use `System.lineSeparator `instead

File: src/main/java/core/basesyntax/service/impl/WriterServiceImp.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    private static final String EMPTY_MESSAGE = "Your report is empty";
+    private static final String WRITE_EXCEPTION = "Can not write report";
+
+    @Override
+    public boolean writeCvsFile(String report, String filePath) {
+        if (report == null || filePath == null) {
+            throw new RuntimeException(NULL_MESSAGE);
+        }
+        if (report.isEmpty()) {
+            throw new RuntimeException(EMPTY_MESSAGE);
```
Comment (kshuryhin) [inline src/main/java/core/basesyntax/service/impl/WriterServiceImp.java:19]: Exception type could be more specific I guess

File: src/main/java/core/basesyntax/strategy/impl/PurchaseOperationHandler.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+        if (Storage.shopStorage.containsKey(fruitName)) {
+            Integer availableQuantity = Storage.shopStorage.get(fruitName);
+            if (availableQuantity < transactionQuantity) {
+                throw new RuntimeException(NOT_ENOUGH_AMOUNT_MESSAGE + fruitName);
+            } else {
+                Storage.shopStorage.put(fruitName, availableQuantity - transactionQuantity);
+            }
+        } else {
+            throw new RuntimeException(STORAGE_FRUIT_EXCEPTION_MESSAGE + fruitName);
+        }
```
Comment (kshuryhin) [inline src/main/java/core/basesyntax/strategy/impl/PurchaseOperationHandler.java:25]: That's cool you've considered so many edge cases, but that will definitely complicate your next task related to tests creation. If you feel ok with that, no changes required. Otherwise, feel free to remove this check

`if (availableQuantity < transactionQuantity) `

**Review by kshuryhin — CHANGES_REQUESTED** [review]
Great job, but there are some points to consider

**Review by mnyshenko — APPROVED** [review]
Approved, great job!
